storage/replica.go

```diff
@@ -2594,6 +2535,78 @@ func (r *Replica) processRaftCommand(
 	return pErr
 }
 
+func (r *Replica) maybeAcquireSplitMergeLock(ba roachpb.BatchRequest) func(pErr *roachpb.Error) {
+	arg, ok := ba.GetArg(roachpb.EndTransaction)
+	if !ok {
+		return nil
+	}
+	ict := arg.(*roachpb.EndTransactionRequest).InternalCommitTrigger
+	if split := ict.GetSplitTrigger(); split != nil {
+		return r.acquireSplitLock(split)
+	}
+	if merge := ict.GetMergeTrigger(); merge != nil {
+		return r.acquireMergeLock(merge)
+	}
+	return nil
+}
+
+func (r *Replica) acquireSplitLock(split *roachpb.SplitTrigger) func(pErr *roachpb.Error) {
+	rightRng, created, err := r.store.getOrCreateReplica(split.RightDesc.RangeID, 0, nil)
+	if err != nil {
+		return nil
+	}
+
+	// It would be nice to assert that rightRng is not initialized
+	// here. Unfortunately, due to reproposals and retries we might be executing
+	// a reproposal for a split trigger that was already executed via a
+	// retry. The reproposed command will not succeed (the transaction has
+	// already committed).
+	//
+	// TODO(peter): It might be okay to return an error here, but it is more
+	// conservative to hit the exact same error paths that we would hit for other
+	// commands that have reproposals interacting with retries (i.e. we don't
+	// treat splits differently).
```

Comment:
I replaced the `rightRng.IsInitialized()` check with the above comment. I think this PR is ready to go. PTAL.